import matplotlib.pyplot as plt
import numpy as np

plt.rcParams["font.family"] = "serif"

mappings = {'C09_FS841':0.85074292,'K20_FS841':0.521733092,'N22_FS841': 0.776108175,'M21_FS841':0.005984661,'E23_FS841':0.578942662,
	'C09_FS842':2.687491353,'K20_FS842':2.687537325,'N22_FS842':13.70974501,'M21_FS842':1.714508436,'E23_FS842':4.885482009,
	'C09_FS844':0.087377417,'K20_FS844':0.034499176,'N22_FS844':0.026152219,'M21_FS844':0.108001639,'E23_FS844':0.024126383,
	'C09_FS848':108.2655521,'K20_FS848':127.1934995,'N22_FS848':34.96652093,'M21_FS848':116.9982544,'E23_FS848':128.0653937,
	'C09_FS849':0.266736873,'K20_FS849':0.359053378,'N22_FS849':0.626042108,'M21_FS849':0.050728756,'E23_FS849':0.23392349,
	'C09_FS851':1.705616837,'K20_FS851':3.396945504,'N22_FS851':4.520771031,'M21_FS851':0.067383183,'E23_FS851':2.324827237,
	'C09_FS852':3.966187332,'K20_FS852':8.987104562,'N22_FS852':11.8071716,'M21_FS852':0.080472645,'E23_FS852':5.298544029,
	'C09_FS854':5.1191866,'K20_FS854':5.79856688,'N22_FS854':4.743035247,'M21_FS854':5.729131702,'E23_FS854':1.072089439,
	'C09_FS856':1.524237066,'K20_FS856':1.634765377,'N22_FS856':1.000254865,'M21_FS856':1.783870947,'E23_FS856':0.210610362,
	'C09_FS866':14.39162359,'K20_FS866':19.1174054,'N22_FS866':69.23744747,'M21_FS866':12.33293687,'E23_FS866':288.0378849,
	'C09_FS872':0.11233089,'K20_FS872':0.009309417,'N22_FS872':0.013612907,'M21_FS872':0.010954871,'E23_FS872':0.020586681,
	'C09_FS874':0.197947084,'K20_FS874':0.010388171,'N22_FS874':0.071649855,'M21_FS874':0.01218419,'E23_FS874':0.060701726,
	'C09_FS877':79.29027457,'K20_FS877':95.06176238,'N22_FS877':18.12472016,'M21_FS877':91.7939652,'E23_FS877':51.7186239,
	'C09_FS879':3.342027416,'K20_FS879':4.233221034,'N22_FS879':7.233275936,'M21_FS879':3.452026452,'E23_FS879':1.957391507,
	'C09_FS881':13.2611894127,'K20_FS881':41.8011974258,'N22_FS881':42.26812025,'M21_FS881':8.45602434432,'E23_FS881':434.018177602}

metagenome_reads = { #number of reads in each metagenome, bp
	'FS841':53842644,
	'FS842':64300743,
	'FS844':82138820,
	'FS848':77333359,
	'FS849':26061468,
	'FS851':75912423,
	'FS852':37753638,
	'FS854':140664299,
	'FS856':104123722,
	'FS866':24088594,
	'FS872':40397883,
	'FS874':65834175,
	'FS877':152063932,
	'FS879':24723930,
	'FS881':167339253
				   }

#non-normalized
c = (0.85074292, 2.687491353, 0.087377417, 108.2655521, 0.266736873, 1.705616837, 3.966187332, 5.1191866, 1.524237066, 14.39162359, 0.11233089, 0.197947084, 79.29027457, 3.342027416, 13.2611894127)
e = (0.578942662, 4.885482009, 0.024126383, 128.0653937, 0.23392349, 2.324827237, 5.298544029, 1.072089439, 0.210610362, 288.0378849, 0.020586681, 0.060701726, 51.7186239, 1.957391507, 434.018177602)
k = (0.521733092, 2.687537325, 0.034499176, 127.1934995, 0.359053378, 3.396945504, 8.987104562, 5.79856688, 1.634765377, 19.1174054, 0.009309417, 0.010388171, 95.06176238, 4.233221034, 41.8011974258)
m = (0.005984661, 1.714508436, 0.108001639, 116.9982544, 0.050728756, 0.067383183, 0.080472645, 5.729131702, 1.783870947, 12.33293687, 0.010954871, 0.01218419, 91.7939652, 3.452026452, 8.45602434432)
n = (0.776108175, 13.70974501, 0.026152219, 34.96652093, 0.626042108, 4.520771031, 11.8071716, 4.743035247, 1.000254865, 69.23744747, 0.013612907, 0.071649855, 18.12472016, 7.233275936, 42.26812025)

#normalized by genome length
c = (4.886827542234145e-07, 1.5437456433203705e-06, 5.019123379420871e-08, 6.218977195572379e-05, 1.5321868297250256e-07, 9.797384309175192e-07, 2.2782526937371115e-06, 2.9405571862667944e-06, 8.755504747571644e-07, 8.266819609510078e-06, 6.452497860289852e-08, 1.1370453273900131e-07, 4.5545826887396674e-05, 1.9197234839650985e-06, 7.617476930018554e-06)
e = (5.545831516470547e-07, 4.679921152306107e-06, 2.3111244688310645e-08, 0.0001226769321329905, 2.2408095800077402e-07, 2.227008131818025e-06, 5.075603232610803e-06, 1.0269803539338113e-06, 2.017487499089972e-07, 0.00027591844320084603, 1.9720478693851275e-08, 5.814764867940578e-08, 4.95425183250888e-05, 1.8750325761262315e-06, 0.0004157564895547952)
k = (2.771755556972247e-06, 1.4277791900420757e-05, 1.8328007990139828e-07, 0.0006757272913213481, 1.907504451952909e-06, 1.8046588805304094e-05, 4.7744828520124106e-05, 3.0805425644948785e-05, 8.684843050065876e-06, 0.00010156299353988695, 4.945714331250797e-08, 5.518812422967402e-08, 0.000505024450571635, 2.2489380307280376e-05, 0.00022207274759764544)
m = (6.419608751757849e-09, 1.8391139215250894e-06, 1.1585088393955679e-07, 0.0001255013471751585, 5.4415574413218815e-08, 7.228039671889531e-08, 8.632116273905949e-08, 6.145508327728596e-06, 1.913517498045046e-06, 1.3229258844490784e-05, 1.175103915593185e-08, 1.306970148469236e-08, 9.84652835568256e-05, 3.7029096923883906e-06, 9.070583594605292e-06)
n = (9.059751883780517e-07, 1.6003811347625358e-05, 3.0528297856196486e-08, 4.081750638238264e-05, 7.307984054257549e-07, 5.277236496605985e-06, 1.3782878288225715e-05, 5.536692422270342e-06, 1.1676285844359748e-06, 8.082302381939055e-05, 1.5890769329543398e-08, 8.363910208893895e-08, 2.115754902498672e-05, 8.443627756798245e-06, 4.934089014830942e-05)

#normalized by number of reads in metagenome
c = (1.580054129585464e-08, 4.1795650059595736e-08, 1.0637773588663678e-09, 1.3999851228497652e-06, 1.0234913589671925e-08, 2.2468217580144952e-08, 1.0505444089917904e-07, 3.6392934357850105e-08, 1.4638710917383457e-08, 5.974455624101598e-07, 2.78061328114644e-09, 3.0067527086046113e-09, 5.214272281871548e-07, 1.3517379381028827e-07, 7.924733243968766e-08)
e = (1.0752493172512108e-08, 7.59786245238255e-08, 2.937269247354661e-10, 1.6560174723562699e-06, 8.975837048012798e-09, 3.0625122280710236e-08, 1.4034525703191836e-07, 7.621617188025797e-09, 2.0226933685678274e-09, 1.1957438649179773e-05, 5.095980153217434e-10, 9.220397460741324e-10, 3.401110521066889e-07, 7.916991784882096e-08, 2.5936423751216336e-06)
k = (9.689960470737655e-09, 4.179636501245406e-08, 4.200106113041312e-10, 1.6447429821327171e-06, 1.3777173948911857e-08, 4.474821603309909e-08, 2.3804605431667277e-07, 4.122273328216707e-08, 1.5700220330195263e-08, 7.936289432251629e-07, 2.304431893126677e-10, 1.5779298517829076e-10, 6.25143392846109e-07, 1.7121958499316248e-07, 2.497991157269e-07)
m = (1.1115094942217176e-10, 2.666389774065286e-08, 1.3148671846028466e-09, 1.5129079599400305e-06, 1.946504164692488e-09, 8.876436864622277e-10, 2.131520278919875e-09, 4.072910996414236e-08, 1.713222417270101e-08, 5.119824291114707e-07, 2.7117438307348927e-10, 1.8507393766231595e-10, 6.03653765838437e-07, 1.396228856820093e-07, 5.053222237295394e-08)
n = (1.4414377105997988e-08, 2.1321285525425422e-07, 3.183904882003418e-10, 4.5215313781986376e-07, 2.4021751499186465e-08, 5.9552453371169564e-08, 3.12742618340516e-07, 3.371882759675929e-08, 9.606407125938122e-09, 2.874283466689671e-06, 3.369708011679721e-10, 1.0883383136494076e-09, 1.1919144744987916e-07, 2.9256173820262396e-07, 2.525893924601181e-07)

labels = ('FS841','FS842','FS844','FS848','FS849','FS851','FS852','FS854','FS856','FS866','FS872','FS874','FS877','FS879','FS881')

def make_plot():
	n_groups = 15

	# create plot
	fig, ax = plt.subplots()
	index = np.arange(n_groups)
	bar_width = 0.15
	opacity = 0.8
 
	rectsc = plt.bar(index, c, bar_width,
					 alpha=opacity,
					 color='r',
					 label='C09')
 
	rectse = plt.bar(index + bar_width, e, bar_width,
					 alpha=opacity,
					 color='b',
					 label='E23')
				 
	rectse = plt.bar(index + (2 * bar_width), k, bar_width,
					 alpha=opacity,
					 color='g',
					 label='K20')
				 
	rectse = plt.bar(index + (3 * bar_width), m, bar_width,
					 alpha=opacity,
					 color='k',
					 label='M21')
				 
	rectse = plt.bar(index + (4 * bar_width), n, bar_width,
					 alpha=opacity,
					 color='y',
					 label='N22')
 
	plt.xlabel('Sample Site')
	plt.ylabel('Normalized Coverage')
	plt.title('Abundance by Sample site')
	plt.xticks(index + (2*bar_width), labels)
	plt.legend()
 
	plt.tight_layout()
	plt.savefig('Abundance_by_sample_site.svg')
	plt.show()



def make_tuples():

	c = ()
	e = ()
	k = ()
	m = ()
	n = ()

	for SAG in ['C09','E23','K20','M21','N22']:
		count = 0
		for sample_site in ['FS841','FS842','FS844','FS848','FS849','FS851','FS852','FS854','FS856','FS866','FS872','FS874','FS877','FS879','FS881']:
			if SAG == 'C09':
				c += (mappings[SAG +'_'+ sample_site] / float(metagenome_reads[sample_site]),)
				count += 1
			
			if SAG == 'E23':
				e += (mappings[SAG + '_'+ sample_site] / float(metagenome_reads[sample_site]),)
				count += 1
		
			if SAG == 'K20':
				k += (mappings[SAG + '_'+ sample_site] / float(metagenome_reads[sample_site]),)
				count += 1
		
			if SAG == 'M21':
				m += (mappings[SAG + '_'+ sample_site] / float(metagenome_reads[sample_site]),)
				count += 1 
			
			if SAG == 'N22':
				n += (mappings[SAG + '_'+ sample_site] / float(metagenome_reads[sample_site]),)
				count += 1

	print(c)
	print(e)
	print(k)
	print(m)
	print(n)
	
	
if __name__ == '__main__':
	make_tuples()
	make_plot()
			
		
		